shared/domains/chelonia/GIMessage.js

```diff
@@ -143,6 +164,17 @@ export class GIMessage {
   serialize (): string { return this._mapping.value }
 
   hash (): string { return this._mapping.key }
+
+  id (): string {
+    const data: string = JSON.stringify({
+      contractID: this.contractID(),
+      op: this.op(),
+      manifest: this.manifest(),
+      nonce: this._message.nonce
+    })
+
+    return blake32Hash(data)
+  }
```

Comment:
So I get that the idea here is to include everything except for the `previousHEAD`, and  yeah that makes sense, however there is an important JavaScript quirk to be aware of that makes it so that we cannot do it this way. There is a comment on line 60 that discusses this a bit, although it doesn't explain why:

```js
    // NOTE: the JSON strings generated here must be preserved forever.
    //       do not ever regenerate this message using the contructor.
    //       instead store it using serialize() and restore it using
    //       deserialize().
```

The issue is that different implementations of JavaScript engines might generate different strings when serializing JS objects using `JSON.stringify`, and that would lead to different hashes resulting from `blake32Hash`. So to get around this we save the serialized string upon creation and keep a copy of it (instead of regenerating it as needed).

So one way to solve this would be to save this hash upon message creation and include it as part of the data that is signed. However, there is also an easier solution, and that is to just create a unique ID. There is a name for this: [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier):

> While the probability that a UUID will be duplicated is not zero, it is generally considered close enough to zero to be negligible

A UUID is 128-bits, that's twice the size of the value of `Math.random()`. You can add the dependency `uuid` to the project to do this, it is considered a high-quality dependency: https://github.com/uuidjs/uuid

Then you can simply replace the `nonce` with a `uuid`, and add an accessor function to access it. This would be the simplest and easiest approach with minimal changes, and you can do it in a way that is DRY. You would still need a a function to copy the GIMessage though.

Reply:
Very interesting to hear that `JSON.stringify()` could make different strings according to the JavaScript version.
So we can make `id()` function which returns the `this._message.uuid`. This is absolutely a simple and easy approach.

> You would still need a a function to copy the GIMessage though.

I tried to do this but as I noted in the comment (Line 83), I don't think we should make a new static function for this.

Reply:
I think we can make a function which returns only `{ mapping, message }`, not `this` typed object.

Reply:
> There is a comment on line 60 that discusses this a bit, although it doesn't explain why:

> The issue is that different implementations of JavaScript engines might generate different strings when serializing JS objects using `JSON.stringify`, and that would lead to different hashes resulting from `blake32Hash`. So to get around this we save the serialized string upon creation and keep a copy of it (instead of regenerating it as needed).

Should this "why" be added to the comment on line 60 in some form?

Reply:
Yeah probably a good idea, thanks @dotmacro